tests/unit/lms/views/predicates/_lti_launch_test.py

```diff
@@ -51,6 +52,84 @@ def test_when_request_params_are_missing(self, pyramid_request, value, expected)
         assert result is expected
 
 
+class TestBlackboardCopied:
+    @pytest.mark.parametrize(
+        "params",
+        [
+            # If there's already a ModuleItemConfiguration for the resource_link_id
+            # then blackboard_copied is False.
+            dict(
+                resource_link_id_exists=True,
+                resource_link_id_history_exists=True,
+                expected_result=False,
+            ),
+            dict(
+                resource_link_id_exists=True,
+                resource_link_id_history_exists=False,
+                expected_result=False,
+            ),
+            # If there isn't a ModuleItemConfiguration for the
+            # resource_link_id_history then blackboard_copied is False.
+            dict(
+                resource_link_id_exists=False,
+                resource_link_id_history_exists=False,
+                expected_result=False,
+            ),
+            # If there's a ModuleItemConfiguration for resource_link_id_history but
+            # not yet one for resource_link_id then blackboard_copied is True.
+            dict(
+                resource_link_id_exists=False,
+                resource_link_id_history_exists=True,
+                expected_result=True,
+            ),
+        ],
+    )
+    def test_it(self, params, pyramid_request):
```

Comment:
For table like content, I think lists can be easier to grok here:

```python
    @pytest.mark.parametrize(
        "link_exists,history_exists,expected_result",
        [
            param(True, True, False, id="already copied")",
            param(True, False, False, id="normal db configured"),
            param(False, False, False, id="no history"),
            param(False, True, True, id="history copy required")
        ]
    )
    def test_it(self, link_exists, history_exists, expected_result, pyramid_request):
```

Another (possibly better) alternative is to directly encode the logic into the test as a declarative rule:

```python
    @pytest.mark.parametrize("link_exists", [True, False])
    @pytest.mark.parametrize("history_exists", [True, False])
    def test_it(self, link_exists, history_exists, pyramid_request):
        # We only need to copy if there is a history and no link has been created
         expected_result == history_exists and not link_exists
```

This has the benefit of being very compact, and actually describing the rule in play, rather than leaving it to the reader to work backwards from the examples.

Reply:
:+1: The doubled `parametrize` decorators work pretty nicely I think